src/geff/affine.py

```diff
@@ -24,55 +74,31 @@ class Affine(BaseModel):
     where p_out_homo = [p_out; 1] and p_in = p_in_homo[:-1]
 
     Attributes:
-        matrix: Homogeneous transformation matrix as list of lists (ndim+1, ndim+1)
+        matrix (np.ndarray) : square, homogeneous transformation matrix (ndim+1, ndim+1)
     """
 
-    model_config = ConfigDict(arbitrary_types_allowed=True)
-
-    matrix: Any = Field(
-        ..., description="Homogeneous transformation matrix as list of lists (ndim+1, ndim+1)"
+    matrix: Annotated[np.ndarray, TFormMeta] = Field(
+        ...,
+        description="Homogeneous transformation matrix (ndim+1, ndim+1)",
     )
 
-    @model_validator(mode="after")
-    def validate_matrix(self) -> Affine:
```

Comment:
side comment, for the future:   Try not to use `model_validator` in cases where you only want to validate a single field.  Use `field_validator` for that (or, in complex cases, like here, you can [customize validation with `__get_pydantic_core_schema__`](https://docs.pydantic.dev/latest/concepts/types/#as-an-annotation))